### ISSUE TITLE
[Progress Report] Emotion Detection with CNN transfer learning approach

### ISSUE THREAD

**Issue by KatJHuang** [capstone496/SpeechSentiments#5]
This thread is created to track the progress in the CNN/transfer learning research.

---

Found an implementation/tutorial on CNN transfer learning and experimented it with our RAVDESS dataset. With no modification to their implementation, we can achieve an accuracy at 42.4%. 

As a quick test, this accuracy is decent. But further tuning on the model would definitely improve the result. The current model is using all conv/pooling layers of the Inception V3 model, which would suit better for real world images as opposed to our B&W image of wavy lines. These many layers are better at the modeling the complicated structures in real images but would be too much for our purposes. We can try using fewer layers and get better modeling behavior. 

Tutorial link: https://codelabs.developers.google.com/codelabs/tensorflow-for-poets/#0

Test accuracy: 42.4%
Result: 
[train_report_no_mod.pdf](https://github.com/capstone496/SpeechSentiments/files/2426159/train_report_no_mod.pdf)

Dataset: RAVDESS
Data preparation: 
1. Trim silence with [voiced_segment_extractor.py](url
https://github.com/KatJHuang/audio_emotion_analysis/blob/master/voiced_segment_extractor.py)
2. Convert waveform to image representation, in this case, melspectrograms and save them to a directory structure supported by the implementation showed in the tutorial above, with [waveform2img.py](
https://github.com/KatJHuang/audio_emotion_analysis/blob/master/waveform2img.py)
The above scripts can be found in my [experimentation repo](https://github.com/KatJHuang/audio_emotion_analysis/blob/master/).

Step to reproduce (assuming RAVDESS speech dataset is already downloaded): 
1. Trim silence and save the shortened audio file:
`python voiced_segment_extractor.py $aggressive_lvl --mode dir $input_dir --output $output_dir`
2. Convert trimmed audio waveforms into spectrograms and group them by their emotion labels: 
`python waveform2img.py $input_dir $output_dir` <img width="583" alt="image" src="https://user-images.githubusercontent.com/6192987/46179112-c79b0900-c287-11e8-9a14-fa1fabb0d95a.png">
3. Get the transfer learning implementation made by Tensorflow, make NO modification to the python code, and run [training commands](https://codelabs.developers.google.com/codelabs/tensorflow-for-poets/#3) verbatim (except using my own data folders from step 2 of course).

**Comment by KatJHuang** [comment]
Note to self: it's also help to generate a confusion matrix. Flag: add this feature.

**Comment by KatJHuang** [comment]
Experimented with transfer learning on the Inception v3 model (https://www.cs.unc.edu/~wliu/papers/GoogLeNet.pdf). It's got some really nice architecture that takes away the conundrum of choosing sizes for convolution kernels. Validation accuracy is around 50% and training hikes up to almost 90% using up to the second last layer of the model. A bit more decent than the MobileNet experiment.

Then to test the hypothesis that using less layers prevents irrelevant high level feature learning, I conducted another experiment that lumped our custom NNs on top of the first pooling layer. This actually worsened the prediction accuracy. The cross entropy loss was not converging. This can be attributed to a gargantuan number of parameters resulting from the lower-level convolution/pooling layers. Specifically the first pooling layer unravels to around 55k parameters (192 kernels of 17x17 pixels) which is rather demanding for our dataset that's around 1k samples. So this approach leads us into underfitting. 

(The above experiment was nice coding challenge nonetheless. The python script provided by Tensorflow contained some ~1k lines of code so it's cool to get sync'ed with it. Then it turned out I only needed to add two additional line of code to achieve my purpose) 

To seek improvement from other avenues, I ran into a few new papers:
- https://www.astesj.com/publications/ASTESJ_030437.pdf
- https://www.sciencedirect.com/science/article/pii/S089360801730059X

And it caught my attention that apparently the color in which our spectrograms are saved have impact on the accuracy of CNN models. Paper 1 actually found RGB produced higher accuracy than theses colors individually or the greyscale image. This is curious. Will conduct another experiment to verify if this is true.

**Comment by KatJHuang** [comment]
(Sidenote: the discussion in the Inception v3 paper has got lots to say about the sparsity of features and how convolution reduces it. Reminds me of 367, like greatly. Perhaps time to get a good foundation)

**Comment by KatJHuang** [comment]
Meeting minutes from Oct 30
- there is overfitting in training with both the Inception v3 and MobileNet. Train accuracy keeps increasing while validation plateaus. Will look up strategies to prevent overfitting.
- Size of data vs. number of parameters. Need proof for their relationship.

**Comment by rightnknow** [comment]
https://arxiv.org/ftp/arxiv/papers/1707/1707.09917.pdf
Check this paper, I think it worth a read

**Comment by KatJHuang** [comment]
CNN code is on https://github.com/KatJHuang/tensorflow-for-poets-2/tree/prevent-overfit-w-dropout
You can `git clone https://github.com/KatJHuang/tensorflow-for-poets-2.git`, and `git checkout prevent-overfit-w-dropout` to get all the relevant source files. The branch contains scripts to [serialize spectrograms data](https://github.com/KatJHuang/tensorflow-for-poets-2/blob/prevent-overfit-w-dropout/scripts/image2tfrecords.py), [run training](https://github.com/KatJHuang/tensorflow-for-poets-2/blob/prevent-overfit-w-dropout/scripts/retrain.py), and [make a prediction](https://github.com/KatJHuang/tensorflow-for-poets-2/blob/prevent-overfit-w-dropout/scripts/main.py). 

Before you run training, make sure your spectrograms are already serialized into TF records file. You can serialize the labelled spectrogram images using the `image2tfrecords.py` script as follows:
`python image2tfrecords.py $LABELLED_SPECTROGRAM_DIR $SERIALIZATION_OUTPUT_DIR`

-> For obtaining a folder of labelled spectrogram images, please consult [this comment from above](https://github.com/capstone496/SpeechSentiments/issues/5#issue-364684173).

### To run training, use the `retrain.py` script
**Input**
Spectrograms serialized into the TF records format

**Output**
- Trained model saved as a .pb file in the `aud_emo/models/` folder
- Training summaries and whatnot (these are important only if you want to halt the training midways and pick it up later)

**Command**
Assuming your `pwd` is the `tensorflow-for-poets-2/scripts/` folder, where the `retrain.py` script is located, you can run
```
python retrain.py
--bottleneck_dir=aud_emo/bottlenecks
--model_dir=aud_emo/models/
--summaries_dir=aud_emo/training_summaries/inception_v3_standard
--output_graph=aud_emo/retrained_graph.pb
--output_labels=aud_emo/retrained_labels.txt
--architecture=inception_v3
--image_records_dir=$SERIALIZATION_OUTPUT_DIR
--how_many_training_steps=1000
```

And training should start. If you are interested in the internals of the training process, you can delve into the `retrain.py` code. The relevant code for producing result adapted to our purpose is in the [function](https://github.com/KatJHuang/tensorflow-for-poets-2/blob/prevent-overfit-w-dropout/scripts/retrain.py#L993:L1062) where the extra fully connected layer is added.
